### ISSUE TITLE
[Spike] Make UKI systemd sysext solution more robust

### ISSUE THREAD

**Issue by jimmykarily** [kairos-io/kairos#2630]
Instead of [XBOOTLDR](https://github.com/kairos-io/kairos/issues/2287), we decided to go further in the path of [systemd sysext}](https://github.com/kairos-io/kairos/issues/2608).

This ticket is to investigate the limit of this solution and find ways to support extensions as big as possible. We have reasons to believe that the system extension size is limited to a very low size (~500Mb) and we want to see if it's possible to load bigger ones.

**Comment by Itxaka** [comment]
Good test for this.

 - Build uki rootfs
 - Pack the /usr/firmware dir into a sysext
 - Remove /usr/firmware from rootfs
 - Build image, add sysext under the overlay-iso dir so it gets picked by the agent on install
 - Install and reboot
 - systemd-stub will try to measure the sysext and pass it to the kernel in a CPIO format. This usually gets copied intot he initramfs under `/.extra/sysext`
 - This fails under QEMU due to an error allocating pages


Hint: A way of skipping this would be to make so the agent does copy into the /EFI/kairos/active.efi.extra.d/ dir, as stub autoloads them from there. We could just store them under /EFI/kairos/sysext and make immucore move those sysext into `/run/extensions` on boot directly. That would skip the systemd-stub loading the files into memory to measure them. As we currently are NOT using those measurements for anything, it would be a quick workaround.

**Comment by Itxaka** [comment]
FYI I can confirm that the workaround works.

Manually copied the firmware.raw extension into /run/extensions and refreshed and magically all the firmware appeared correctly :D

**Comment by Itxaka** [comment]
lol, nice, this works on systemd-256 I was able to auto copy the firmware sysext and have it autoload on boot :D

**Comment by jimmykarily** [comment]
Before we close this, let's bump the sizes until things break, just to know what our limit is.

**Comment by mudler** [comment]
Depends on #2595 https://github.com/kairos-io/packages/pull/916 https://github.com/kairos-io/packages/pull/919

**Comment by jimmykarily** [comment]
> Tumbleweed is a bit iffy if I recall my tests. I would recommend not testing in there. Even fedora 40 has some bugs so Ubuntu is the safe choice for uki

I was trying to get a systemd with systemd 256 but then I realised we now have a packages :facepalm: I will try again with ubuntu, I'm not sure how I ended up with system 255. I probably used the wrong image to build the uki iso.

**Comment by jimmykarily** [comment]
> > Tumbleweed is a bit iffy if I recall my tests. I would recommend not testing in there. Even fedora 40 has some bugs so Ubuntu is the safe choice for uki
> 
> I was trying to get a system with systemd 256 but then I realised we now have a packages 🤦 I will try again with ubuntu, I'm not sure how I ended up with system 255. I probably used the wrong image to build the uki iso.

I ended up with 255 because we don't consume the package yet. I will do hacks to consume it and see.

**Comment by jimmykarily** [comment]
> > > Tumbleweed is a bit iffy if I recall my tests. I would recommend not testing in there. Even fedora 40 has some bugs so Ubuntu is the safe choice for uki
> > 
> > 
> > I was trying to get a system with systemd 256 but then I realised we now have a packages 🤦 I will try again with ubuntu, I'm not sure how I ended up with system 255. I probably used the wrong image to build the uki iso.
> 
> I ended up with 255 because we don't consume the package yet. I will do hacks to consume it and see.

nope, [we just build the systemd-boot files](https://github.com/kairos-io/packages/blob/main/packages/system/systemd/build.yaml). @Itxaka what part of systemd 256 was expected to improve things and in which way? The only reliable flavor for uki isos is Ubuntu and that one doesn't have systemd 256, so I'm wondering, why is version 256 considered important here?

My tests so far have shown that with systemd 255 in qemu, we hit the firmware limit somewhere around 500Mb.

**Comment by jimmykarily** [comment]
> > > > > Tumbleweed is a bit iffy if I recall my tests. I would recommend not testing in there. Even fedora 40 has some bugs so Ubuntu is the safe choice for uki
> > > > 
> > > > 
> > > > I was trying to get a system with systemd 256 but then I realised we now have a packages 🤦 I will try again with ubuntu, I'm not sure how I ended up with system 255. I probably used the wrong image to build the uki iso.
> > > 
> > > 
> > > I ended up with 255 because we don't consume the package yet. I will do hacks to consume it and see.
> > 
> > 
> > nope, [we just build the systemd-boot files](https://github.com/kairos-io/packages/blob/main/packages/system/systemd/build.yaml). @Itxaka what part of systemd 256 was expected to improve things and in which way? The only reliable flavor for uki isos is Ubuntu and that one doesn't have systemd 256, so I'm wondering, why is version 256 considered important here?
> > My tests so far have shown that with systemd 255 in qemu, we hit the firmware limit somewhere around 500Mb.
> 
> The loading I guess but maybe I screwed up and mixed versions? Or increased ram somehow and that made it work?
> 
> Maybe we should just ignore the system version and bump memory to see if the loading is linear (i.e. with double the ram we can load double size sysext) so we can write that down?

My example above shows that 500Mb sysext fit in 8G of RAM but 1G of sysext couldn't run in a 18Gb RAM vm, so it's not linear.

**Comment by Itxaka** [comment]
> > > > > > Tumbleweed is a bit iffy if I recall my tests. I would recommend not testing in there. Even fedora 40 has some bugs so Ubuntu is the safe choice for uki
> > > > > 
> > > > > 
> > > > > I was trying to get a system with systemd 256 but then I realised we now have a packages 🤦 I will try again with ubuntu, I'm not sure how I ended up with system 255. I probably used the wrong image to build the uki iso.
> > > > 
> > > > 
> > > > I ended up with 255 because we don't consume the package yet. I will do hacks to consume it and see.
> > > 
> > > 
> > > nope, [we just build the systemd-boot files](https://github.com/kairos-io/packages/blob/main/packages/system/systemd/build.yaml). @Itxaka what part of systemd 256 was expected to improve things and in which way? The only reliable flavor for uki isos is Ubuntu and that one doesn't have systemd 256, so I'm wondering, why is version 256 considered important here?
> > > My tests so far have shown that with systemd 255 in qemu, we hit the firmware limit somewhere around 500Mb.
> > 
> > The loading I guess but maybe I screwed up and mixed versions? Or increased ram somehow and that made it work?
> > 
> > Maybe we should just ignore the system version and bump memory to see if the loading is linear (i.e. with double the ram we can load double size sysext) so we can write that down?
> 
> My example above shows that 500Mb sysext fit in 8G of RAM but 1G of sysext couldn't run in a 18Gb RAM vm, so it's not linear.

Oh well, that sucks. So maybe there is a max allocated to the EFI implementation that we cant go over.

Did you try a 800mb one? IIRC we had the same with EFI files, a max of 1Gb so it makes sense that it would be similar

**Comment by jimmykarily** [comment]
> > > > > > > Tumbleweed is a bit iffy if I recall my tests. I would recommend not testing in there. Even fedora 40 has some bugs so Ubuntu is the safe choice for uki
> > > > > > 
> > > > > > 
> > > > > > I was trying to get a system with systemd 256 but then I realised we now have a packages 🤦 I will try again with ubuntu, I'm not sure how I ended up with system 255. I probably used the wrong image to build the uki iso.
> > > > > 
> > > > > 
> > > > > I ended up with 255 because we don't consume the package yet. I will do hacks to consume it and see.
> > > > 
> > > > 
> > > > nope, [we just build the systemd-boot files](https://github.com/kairos-io/packages/blob/main/packages/system/systemd/build.yaml). @Itxaka what part of systemd 256 was expected to improve things and in which way? The only reliable flavor for uki isos is Ubuntu and that one doesn't have systemd 256, so I'm wondering, why is version 256 considered important here?
> > > > My tests so far have shown that with systemd 255 in qemu, we hit the firmware limit somewhere around 500Mb.
> > > 
> > > 
> > > The loading I guess but maybe I screwed up and mixed versions? Or increased ram somehow and that made it work?
> > > Maybe we should just ignore the system version and bump memory to see if the loading is linear (i.e. with double the ram we can load double size sysext) so we can write that down?
> > 
> > 
> > My example above shows that 500Mb sysext fit in 8G of RAM but 1G of sysext couldn't run in a 18Gb RAM vm, so it's not linear.
> 
> Oh well, that sucks. So maybe there is a max allocated to the EFI implementation that we cant go over.
> 
> Did you try a 800mb one? IIRC we had the same with EFI files, a max of 1Gb so it makes sense that it would be similar

I will try more sizes and record the findings here. In any case, these are all based on qemu so the actual numbers don't really matter. If there is a limit, that limit comes from the firmware so it's value will vary. We can only gather some data points here to know what to expect (e.g. limit is the same and the OS image?). Let's see.

**Comment by jimmykarily** [comment]
800Mb sysext loads fine (18Gb RAM):

```
kairos@localhost:~$ du -h /usr/local/bin/bigfile
800M	/usr/local/bin/bigfile
```

**Comment by jimmykarily** [comment]
Another test:

I created 2 system extensions 700Mb each. As shown above, each one would otherwise load fine. I gave the VM 10G of RAM (although 8G was enough to load a sysext of 800Mb).
The VM didn't boot (got the regular error).

So they extension size adds up. Splitting the files we want into separate extensions won't do the trick.

**Comment by jimmykarily** [comment]
closing as spike is done
